File: tenet-0/shared/bus-ts/src/bus.ts
import { Pool, Client } from "pg";
import {
  ErrConnectionLost,
  ErrCausalityLoop,
  ErrConstitutionRejected,
  ErrNamespaceViolation,
  ErrNoConstitution,
  ErrUnauthenticated,
} from "./errors.js";
import {
  matchesPattern,
  parsePattern,
  patternToLike,
  type ParsedPattern,
} from "./patterns.js";
import { budgetStatus, spStatus } from "./status.js";
import type {
  BudgetStatusResult,
  Config,
  Event,
  PublishOptions,
  SubscriptionHandler,
} from "./types.js";

// Bus is a connected Tenet-0 client. Construct via Bus.connect().
export class Bus {
  readonly config: Config;
  readonly pool: Pool;
  private closed = false;
  private subs: Subscription[] = [];

  private constructor(config: Config, pool: Pool) {
    this.config = config;
    this.pool = pool;
  }

  static async connect(config: Config): Promise<Bus> {
    const pool = new Pool({ connectionString: config.postgresUrl });
    try {
      const { rows } = await pool.query(
        `SELECT status, limit_cents, spent_cents, remaining_cents FROM check_budget($1)`,
        [config.credential],
      );
      const row = rows[0];
      if (!row || row.status === budgetStatus.unauthenticated) {
        throw new ErrUnauthenticated();
      }
      return new Bus(config, pool);
    } catch (err) {
      await pool.end().catch(() => {});
      throw err;
    }
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    const subs = this.subs;
    this.subs = [];
    await Promise.all(subs.map((s) => s.close()));
    await this.pool.end();
  }

  isClosed(): boolean {
    return this.closed;
  }

  async publish(
    eventType: string,
    payload: string | Buffer | null | undefined,
    opts: PublishOptions = {},
  ): Promise<string> {
    // Normalize empty/null payload to "{}" — events.payload is NOT NULL.
    let body: string;
    if (payload == null) {
      body = "{}";
    } else if (Buffer.isBuffer(payload)) {
      body = payload.length === 0 ? "{}" : payload.toString("utf8");
    } else {
      body = payload.length === 0 ? "{}" : payload;
    }

    const { rows } = await this.pool.query(
      `SELECT status, event_id, error_msg FROM publish_event($1, $2, $3::jsonb, $4, $5)`,
      [
        this.config.credential,
        eventType,
        body,
        opts.parentEventId ?? null,
        opts.approvalEventId ?? null,
      ],
    );
    const row = rows[0];
    if (!row) throw new Error("bus: publish returned no row");
    const { status, event_id, error_msg } = row as {
      status: string | null;
      event_id: string | null;
      error_msg: string | null;
    };
    if (status == null) throw new Error("bus: publish returned null status");

    switch (status) {
      case spStatus.ok:
        if (event_id == null) {
          throw new Error("bus: ok status but null event_id");
        }
        return event_id;
      case spStatus.rejectedUnauthenticated:
        throw new ErrUnauthenticated();
      case spStatus.rejectedNamespace:
        throw new ErrNamespaceViolation();
      case spStatus.rejectedConstitution:
        throw new ErrConstitutionRejected(error_msg ?? undefined);
      case spStatus.rejectedCausality:
        throw new ErrCausalityLoop();
      case spStatus.rejectedNoConstitution:
        throw new ErrNoConstitution();
      default:
        throw new Error(
          `bus: publish rejected with status=${status}${error_msg ? `: ${error_msg}` : ""}`,
        );
    }
  }

  async subscribe(
    key: string,
    pattern: string,
    handler: SubscriptionHandler,
  ): Promise<Subscription> {
    if (this.closed) throw new ErrConnectionLost();

    await this.pool.query(`SELECT register_subscription($1, $2, $3)`, [
      this.config.credential,
      key,
      pattern,
    ]);

    const sub = new Subscription(this, key, parsePattern(pattern), handler);
    this.subs.push(sub);
    await sub.start();
    return sub;
  }

  async checkBudget(): Promise<BudgetStatusResult> {
    const { rows } = await this.pool.query(
      `SELECT status, limit_cents, spent_cents, remaining_cents FROM check_budget($1)`,
      [this.config.credential],
    );
    const row = rows[0];
    if (!row) throw new ErrUnauthenticated();
    if (row.status === budgetStatus.unauthenticated) {
      throw new ErrUnauthenticated();
    }
    return {
      status: row.status,
      limitCents: Number(row.limit_cents),
      spentCents: Number(row.spent_cents),
      remainingCents: Number(row.remaining_cents),
    };
  }
}

// Subscription streams events matching a pattern. close() to stop.
//
// Delivery ordering: the dedicated LISTEN client emits notifications
// asynchronously, but we chain each deliver through `deliveryQueue` so
// handlers run serially in notification order — matching the Go
// WaitForNotification loop. At-least-once semantics (same as Go).
export class Subscription {
  private readonly bus: Bus;
  private readonly key: string;
  private readonly pattern: ParsedPattern;
  private readonly handler: SubscriptionHandler;
  private listenClient: Client | null = null;
  private closed = false;
  private deliveryQueue: Promise<void> = Promise.resolve();

  constructor(
    bus: Bus,
    key: string,
    pattern: ParsedPattern,
    handler: SubscriptionHandler,
  ) {
    this.bus = bus;
    this.key = key;
    this.pattern = pattern;
    this.handler = handler;
  }

  async start(): Promise<void> {
    await this.replayMissed();

    // Dedicated client for LISTEN so a long-held listener doesn't pin a pool slot.
    const client = new Client({ connectionString: this.bus.config.postgresUrl });
    await client.connect();
    this.listenClient = client;

    client.on("notification", (msg) => {
      if (msg.channel !== "event_bus" || !msg.payload) return;
      const { id, eventType } = parseNotifyPayload(msg.payload);
      if (eventType && !matchesPattern(this.pattern, eventType)) return;
      // Chain onto the delivery queue so handlers run in notification order.
      this.deliveryQueue = this.deliveryQueue.then(() =>
        this.deliverById(id).catch((err) => {
          if (this.closed) return;
          console.error(
            `subscription ${this.key}: deliver ${id} failed:`,
            err,
          );
        }),
      );
    });

    // Connection-level errors tear down the listen client. Full reconnect is
    // deferred (matches Go behavior — caller re-subscribes on Bus reconnect).
    client.on("error", (err) => {
      if (this.closed) return;
      console.error(
        `subscription ${this.key}: listen client error, stopping:`,
        err,
      );
      void this.close();
    });

    await client.query("LISTEN event_bus");
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    if (this.listenClient) {
      try {
        await this.listenClient.query("UNLISTEN event_bus");
      } catch {
        // Ignore — connection may already be gone.
      }
      await this.listenClient.end().catch(() => {});
      this.listenClient = null;
    }
    // Drain any in-flight deliveries so close() semantics mirror Go's wait.
    await this.deliveryQueue;
  }

  private async replayMissed(): Promise<void> {
    const { rows: cursorRows } = await this.bus.pool.query(
      `SELECT e.published_at
         FROM event_subscriptions es
         LEFT JOIN events e ON e.id = es.last_consumed_event_id
        WHERE es.department_id = $1 AND es.subscription_key = $2`,
      [this.bus.config.department, this.key],
    );
    const lastAt: Date | null = cursorRows[0]?.published_at ?? null;

    const like = patternToLike(this.pattern);
    const query =
      lastAt == null
        ? {
            text: `SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
                     FROM events WHERE event_type LIKE $1 ORDER BY published_at ASC`,
            values: [like],
          }
        : {
            text: `SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
                     FROM events WHERE event_type LIKE $1 AND published_at > $2
                    ORDER BY published_at ASC`,
            values: [like, lastAt],
          };

    const { rows } = await this.bus.pool.query(query.text, query.values);
    for (const row of rows) {
      await this.deliverEvent(rowToEvent(row));
    }
  }

  private async deliverById(eventId: string): Promise<void> {
    const { rows } = await this.bus.pool.query(
      `SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
         FROM events WHERE id = $1`,
      [eventId],
    );
    const row = rows[0];
    if (!row) return;
    const ev = rowToEvent(row);
    if (!matchesPattern(this.pattern, ev.type)) return;
    await this.deliverEvent(ev);
  }

  private async deliverEvent(ev: Event): Promise<void> {
    try {
      await this.handler(ev);
    } catch (err) {
      console.error(
        `subscription ${this.key}: handler error on ${ev.id}:`,
        err,
      );
      return; // No ack → SP will redeliver on next replay.
    }
    await this.bus.pool.query(`SELECT ack_event($1, $2, $3)`, [
      this.bus.config.credential,
      this.key,
      ev.id,
    ]);
  }
}

function rowToEvent(row: Record<string, unknown>): Event {
  const payload = row.payload;
  const payloadStr =
    typeof payload === "string"
      ? payload
      : payload == null
        ? "{}"
        : JSON.stringify(payload);
  return {
    id: String(row.id),
    type: String(row.event_type),
    source: String(row.source_department_id),
    payload: payloadStr,
    parentId: row.parent_event_id == null ? "" : String(row.parent_event_id),
    publishedAt: row.published_at as Date,
  };
}

function parseNotifyPayload(payload: string): { id: string; eventType: string } {
  const idx = payload.indexOf(":");
  if (idx >= 0) {
    return { id: payload.slice(0, idx), eventType: payload.slice(idx + 1) };
  }
  return { id: payload, eventType: "" };
}

File: tenet-0/shared/bus-go/bus.go
package bus

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// Bus is a connected Tenet-0 client. Zero value is not usable — call Connect.
type Bus struct {
	config Config
	pool   *pgxpool.Pool
	logger *slog.Logger

	mu sync.Mutex
	// closed guards Close() idempotency and prevents Subscribe-after-Close.
	closed bool
	subs   []*Subscription
}

// Connect opens a pool against config.PostgresURL and verifies the credential
// via a cheap check_budget() call. Returns ErrUnauthenticated for bad creds.
func Connect(ctx context.Context, config Config) (*Bus, error) {
	pool, err := pgxpool.New(ctx, config.PostgresURL)
	if err != nil {
		return nil, fmt.Errorf("bus: connect pool: %w", err)
	}

	var status string
	var limit, spent, remaining int
	err = pool.QueryRow(ctx,
		`SELECT status, limit_cents, spent_cents, remaining_cents FROM check_budget($1)`,
		config.Credential,
	).Scan(&status, &limit, &spent, &remaining)
	if err != nil {
		pool.Close()
		return nil, fmt.Errorf("bus: credential check: %w", err)
	}
	if status == budgetStatusUnauthenticated {
		pool.Close()
		return nil, ErrUnauthenticated
	}

	return &Bus{
		config: config,
		pool:   pool,
		logger: slog.Default().With("bus.department", config.Department),
	}, nil
}

// Close shuts down all subscriptions and closes the underlying pool.
// Subsequent Subscribe calls return ErrConnectionLost. Idempotent.
func (b *Bus) Close() {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return
	}
	b.closed = true
	subs := b.subs
	b.subs = nil
	b.mu.Unlock()

	for _, s := range subs {
		s.Close()
	}
	b.pool.Close()
}

// Publish writes an event via the publish_event stored procedure. On
// constitutional or authorization rejection returns a typed error.
func (b *Bus) Publish(ctx context.Context, eventType string, payload json.RawMessage, opts ...PublishOption) (string, error) {
	var o publishOptions
	for _, opt := range opts {
		opt(&o)
	}

	var parentArg, approvalArg any
	if o.parentEventID != "" {
		parentArg = o.parentEventID
	}
	if o.approvalEventID != "" {
		approvalArg = o.approvalEventID
	}

	// Normalize nil payload to "{}" so callers don't need to specify it for
	// metadata-only events. The events.payload column is NOT NULL.
	body := []byte(payload)
	if len(body) == 0 {
		body = []byte("{}")
	}

	var status, eventID, errMsg *string
	err := b.pool.QueryRow(ctx,
		`SELECT status, event_id, error_msg FROM publish_event($1, $2, $3, $4, $5)`,
		b.config.Credential, eventType, body, parentArg, approvalArg,
	).Scan(&status, &eventID, &errMsg)
	if err != nil {
		return "", fmt.Errorf("bus: publish: %w", err)
	}

	if status == nil {
		return "", errors.New("bus: publish returned null status")
	}

	switch *status {
	case spStatusOK:
		if eventID == nil {
			return "", errors.New("bus: ok status but nil event_id")
		}
		return *eventID, nil
	case spStatusRejectedUnauthenticated:
		return "", ErrUnauthenticated
	case spStatusRejectedNamespace:
		return "", ErrNamespaceViolation
	case spStatusRejectedConstitution:
		if errMsg != nil {
			return "", fmt.Errorf("%w: %s", ErrConstitutionRejected, *errMsg)
		}
		return "", ErrConstitutionRejected
	case spStatusRejectedCausality:
		return "", ErrCausalityLoop
	case spStatusRejectedNoConstitution:
		return "", ErrNoConstitution
	default:
		if errMsg != nil {
			return "", fmt.Errorf("bus: publish rejected with status=%s: %s", *status, *errMsg)
		}
		return "", fmt.Errorf("bus: publish rejected with status=%s", *status)
	}
}

// Subscribe registers a subscription and starts a background delivery loop.
// The caller's ctx is used only for the initial register_subscription call —
// the delivery loop uses an independent context so request-scoped
// cancellations do not kill the subscription.
func (b *Bus) Subscribe(ctx context.Context, key, pattern string, handler SubscriptionHandler) (*Subscription, error) {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return nil, ErrConnectionLost
	}
	b.mu.Unlock()

	if _, err := b.pool.Exec(ctx,
		`SELECT register_subscription($1, $2, $3)`,
		b.config.Credential, key, pattern,
	); err != nil {
		return nil, fmt.Errorf("bus: register_subscription: %w", err)
	}

	subCtx, cancel := context.WithCancel(context.Background())
	sub := &Subscription{
		bus:     b,
		key:     key,
		pattern: parsePattern(pattern),
		handler: handler,
		ctx:     subCtx,
		cancel:  cancel,
		done:    make(chan struct{}),
	}

	// Append under lock before spawning so Close() observes this sub.
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		cancel()
		close(sub.done)
		return nil, ErrConnectionLost
	}
	b.subs = append(b.subs, sub)
	b.mu.Unlock()

	go sub.run()

	return sub, nil
}

// Subscription is an active subscription. Close stops delivery.
type Subscription struct {
	bus     *Bus
	key     string
	pattern parsedPattern
	handler SubscriptionHandler

	ctx    context.Context
	cancel context.CancelFunc
	done   chan struct{}
}

// Close stops the subscription and waits for the background goroutine to exit.
func (s *Subscription) Close() {
	s.cancel()
	<-s.done
}

// run is the background delivery loop. Replays missed events, then streams
// new ones via a dedicated LISTEN connection. Uses a fresh pgx.Conn so a
// long-held LISTEN does not starve the shared pool.
func (s *Subscription) run() {
	defer close(s.done)

	if err := s.replayMissed(); err != nil {
		s.bus.logger.Error("subscription: replay failed", "key", s.key, "error", err)
	}

	// Dedicated connection for LISTEN so we don't pin a pool slot.
	listenConn, err := pgx.Connect(s.ctx, s.bus.config.PostgresURL)
	if err != nil {
		s.bus.logger.Error("subscription: connect listen", "key", s.key, "error", err)
		return
	}
	defer func() { _ = listenConn.Close(context.Background()) }()

	if _, err := listenConn.Exec(s.ctx, "LISTEN event_bus"); err != nil {
		s.bus.logger.Error("subscription: LISTEN failed", "key", s.key, "error", err)
		return
	}

	for {
		notif, err := listenConn.WaitForNotification(s.ctx)
		if err != nil {
			if s.ctx.Err() != nil {
				return
			}
			s.bus.logger.Error("subscription: wait for notification", "key", s.key, "error", err)
			return
		}

		eventID, eventType := parseNotifyPayload(notif.Payload)
		// Server-side filter avoids a DB round-trip for mismatched events.
		if eventType != "" && !s.pattern.matches(eventType) {
			continue
		}
		if err := s.deliverByID(eventID); err != nil {
			s.bus.logger.Error("subscription: deliver failed", "key", s.key, "event_id", eventID, "error", err)
		}
	}
}

// parseNotifyPayload splits "<id>:<event_type>". Older SP versions may emit
// just the ID; in that case the caller does a fetch-then-filter fallback.
func parseNotifyPayload(payload string) (id, eventType string) {
	if idx := strings.IndexByte(payload, ':'); idx >= 0 {
		return payload[:idx], payload[idx+1:]
	}
	return payload, ""
}

// replayMissed reads events after the subscription's last consumed event and
// delivers them in order.
func (s *Subscription) replayMissed() error {
	ctx := s.ctx
	var lastAt *time.Time
	err := s.bus.pool.QueryRow(ctx, `
		SELECT e.published_at
		FROM event_subscriptions es
		LEFT JOIN events e ON e.id = es.last_consumed_event_id
		WHERE es.department_id = $1 AND es.subscription_key = $2`,
		s.bus.config.Department, s.key,
	).Scan(&lastAt)
	if err != nil && !errors.Is(err, pgx.ErrNoRows) {
		return fmt.Errorf("replay: lookup offset: %w", err)
	}

	rows, err := s.queryEvents(ctx, lastAt)
	if err != nil {
		return fmt.Errorf("replay: query: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var e Event
		var parent *string
		if err := rows.Scan(&e.ID, &e.Type, &e.Source, &e.Payload, &parent, &e.PublishedAt); err != nil {
			return fmt.Errorf("replay: scan: %w", err)
		}
		if parent != nil {
			e.ParentID = *parent
		}
		if err := s.deliverEvent(e); err != nil {
			s.bus.logger.Error("replay: handler error", "event_id", e.ID, "error", err)
		}
	}
	return rows.Err()
}

// deliverByID fetches an event by ID and delivers it if the pattern matches.
// Pattern mismatch returns nil (skipped, not an error).
func (s *Subscription) deliverByID(eventID string) error {
	row := s.bus.pool.QueryRow(s.ctx,
		`SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
		   FROM events WHERE id = $1`, eventID)
	var e Event
	var parent *string
	if err := row.Scan(&e.ID, &e.Type, &e.Source, &e.Payload, &parent, &e.PublishedAt); err != nil {
		return fmt.Errorf("fetch event %s: %w", eventID, err)
	}
	if parent != nil {
		e.ParentID = *parent
	}
	if !s.pattern.matches(e.Type) {
		return nil
	}
	return s.deliverEvent(e)
}

// deliverEvent invokes the handler and, on success, acks the event. The ack
// uses a fresh context with a short timeout so shutdown-induced cancellation
// doesn't lose acks for successfully handled events.
func (s *Subscription) deliverEvent(e Event) error {
	if err := s.handler(s.ctx, e); err != nil {
		return err
	}
	ackCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	_, err := s.bus.pool.Exec(ackCtx,
		`SELECT ack_event($1, $2, $3)`,
		s.bus.config.Credential, s.key, e.ID)
	return err
}

// queryEvents returns matching events newer than after (may be nil).
func (s *Subscription) queryEvents(ctx context.Context, after *time.Time) (pgx.Rows, error) {
	like := s.pattern.toLike()
	if after == nil {
		return s.bus.pool.Query(ctx,
			`SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
			   FROM events WHERE event_type LIKE $1 ORDER BY published_at ASC`,
			like)
	}
	return s.bus.pool.Query(ctx,
		`SELECT id, event_type, source_department_id, payload, parent_event_id, published_at
		   FROM events WHERE event_type LIKE $1 AND published_at > $2 ORDER BY published_at ASC`,
		like, *after)
}
